ctl/install/install.go

```diff
@@ -0,0 +1,167 @@
+package install
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+	"net/http"
+
+	"github.com/spf13/cobra"
+	"k8s.io/client-go/tools/clientcmd"
+
+	utilyaml "k8s.io/apimachinery/pkg/util/yaml"
+
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/discovery/cached/memory"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+)
+
+type Resource struct {
+	Name string
+	URL  string
+}
+
+var decUnstructured = yaml.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
+
+func doServerSideApply(ctx context.Context, cfg *rest.Config, obj *unstructured.Unstructured) error {
+	dc, err := discovery.NewDiscoveryClientForConfig(cfg)
+	if err != nil {
+		return err
+	}
+	mapper := restmapper.NewDeferredDiscoveryRESTMapper(memory.NewMemCacheClient(dc))
+
+	dyn, err := dynamic.NewForConfig(cfg)
+	if err != nil {
+		return err
+	}
+
+	mapping, err := mapper.RESTMapping(obj.GroupVersionKind().GroupKind(), obj.GroupVersionKind().Version)
+	if err != nil {
+		return err
+	}
+
+	var dr dynamic.ResourceInterface
+	if mapping.Scope.Name() == meta.RESTScopeNameNamespace {
+		dr = dyn.Resource(mapping.Resource).Namespace(obj.GetNamespace())
+	} else {
+		dr = dyn.Resource(mapping.Resource)
+	}
+
+	data, err := json.Marshal(obj)
+	if err != nil {
+		return err
+	}
+
+	_, err = dr.Patch(ctx, obj.GetName(), types.ApplyPatchType, data, metav1.PatchOptions{
+		FieldManager: "sample-controller",
+	})
+
+	return err
+}
+
+func NewCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "install",
+		Short: "install kmesh with all the resources",
+		Run: func(cmd *cobra.Command, args []string) {
+			version, err := cmd.Flags().GetString("version")
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			kubeconfig := os.Getenv("HOME") + "/.kube/config"
+
+			config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+			if err != nil {
+				panic(err.Error())
+			}
+
+			resources := []Resource{
```

Comment:
Why not just apply all files in the `yaml` diretory instead of listing each files which may change in the future?

Reply:
How would that work?

Reply:
You could get all yaml contents and call [this function](https://github.com/istio/istio/blob/master/pkg/kube/client.go#L203)

Reply:
I don't know if there is an elegant way to direclty get the contents of all files under yaml directory or to know which files are in this directory and download them one by one.

Reply:
> I don't know if there is an elegant way to direclty get the contents of all files under yaml directory or to know which files are in this directory and download them one by one.

 curl -s "https://api.github.com/repos/kmesh-net/kmesh/contents/deploy/yaml" | jq -r '.[] | .download_url'

Reply:
The `ApplyYAMLContents` function works fine but it requires namespace argument. For now, I can separate the `kmesh-system` and `istio` but it can break in the future because there could be a new yaml file with a different namespace.

Reply:
I think you could specify namespace as "", so that it will direcly use the ns specified in yaml.

Reply:
Oh, that's helpful. I was parsing the namespace field in every file.